taglib/tagutils.cpp

```diff
@@ -38,11 +38,22 @@ long Utils::findID3v1(File *file)
   if(!file->isValid())
     return -1;
 
-  file->seek(-128, File::End);
-  const long p = file->tell();
+  // Differentiate between a match of APEv2 magic and a match of ID3v1 magic.
 
-  if(file->readBlock(3) == ID3v1::Tag::fileIdentifier())
-    return p;
+  if (file->length() >= 131) {
+    file->seek(-131, File::End);
+    const long p = file->tell() + 3;
+    const TagLib::ByteVector data = file->readBlock(8);
+
+    if(data.containsAt(ID3v1::Tag::fileIdentifier(), 3) && (data != APE::Tag::fileIdentifier()))
+      return p;
+  } else {
+    file->seek(-128, File::End);
+    const long p = file->tell();
+
+    if(file->readBlock(3) == ID3v1::Tag::fileIdentifier())
+      return p;
+  }
```

Comment:
The inefficiency here is that `file->length()` uses two seeks. I would suggest the following code, which only uses additional system calls in the rare case of an ID3v1/APEv2 collision:

```
long Utils::findID3v1(File *file)
{
  if(!file->isValid())
    return -1;

  file->seek(-128, File::End);
  const long p = file->tell();

  const ByteVector data = file->readBlock(5);
  if(data.startsWith(ID3v1::Tag::fileIdentifier())) {
    if(data.endsWith("EX")) {
      // Differentiate between a match of APEv2 magic and a match of ID3v1 magic.
      file->seek(-131, File::End);
      if(file->readBlock(8) == APE::Tag::fileIdentifier())
        return -1;
    }
    return p;
  }

  return -1;
}
```

Reply:
Seek and tell are nearly cost-free. In which case the current patch is preferable because it saves a read over the worst-case, and this would happen a lot with APEv2 tags. The current patch should also perform equivalently to the current code.

Quick test clearing caches between runs, `time bash -c 'for f in *; do tagreader "$f" &>/dev/null; done'` on a HDD directory of 156 files containing APEv2 tags. At least two runs, last run result:

Current code: 
`real    0m9.587s
user    0m1.460s
sys     0m0.876s`

Current patch:
`real    0m8.393s
user    0m1.368s
sys     0m0.775s`

Suggested:
`real    0m8.566s
user    0m1.250s
sys     0m0.730s`

No appreciable difference between either patch. Not sure what caused a full second penalty for the current code. All 3 runs were built with the same compiler flags.

Reply:
OK, you are right, one should always measure rather than assume what is faster. I will merge the current state.